src/handlers/permissions-revoke.ts

```diff
@@ -0,0 +1,115 @@
+import { PermissionsRevoke } from '../interfaces/permissions-revoke.js';
+import { DwnInterfaceName, DwnMethodName, Message } from '../core/message.js';
+
+export class PermissionsRevokeHandler implements MethodHandler {
+  constructor(private didResolver: DidResolver, private messageStore: MessageStore, private eventLog: EventLog) { }
+
+  public async handle({
+    tenant,
+    message
+  }: { tenant: string, message: PermissionsRevokeMessage }): Promise<BaseMessageReply> {
+    let permissionsRevoke: PermissionsRevoke;
+    try {
+      permissionsRevoke = await PermissionsRevoke.parse(message);
+    } catch (e) {
+      return messageReplyFromError(e, 400);
+    }
+
+
+    // Authentication
+    try {
+      await authenticate(message.authorization, this.didResolver);
+    } catch (e) {
+      return messageReplyFromError(e, 401);
+    }
+
+    // Authorization
+    const permissionsGrantId = message.descriptor.permissionsGrantId;
+    const permissionsGrantMessage = await this.messageStore.get(
+      tenant,
+      permissionsGrantId
+    ) as PermissionsGrantMessage | undefined;
+
+    if (permissionsGrantMessage === undefined) {
+      return {
+        status: {
+          code   : 400,
+          detail : `Could not find PermissionsGrant with CID ${permissionsGrantId}`
+        }
+      };
+    }
+
+    try {
+      await permissionsRevoke.authorize(permissionsGrantMessage);
+    } catch (e) {
+      return messageReplyFromError(e, 401);
+    }
+
+    // Revoke must have `dateCreated` after that of the grant
+    if (message.descriptor.dateCreated < permissionsGrantMessage.descriptor.dateCreated) {
+      return {
+        status: { code: 400, detail: 'PermissionsRevoke has earlier date than associated PermissionsGrant' }
+      };
+    }
+
+    // query for previous revocations of this grant
+    const query = {
+      interface : DwnInterfaceName.Permissions,
+      method    : DwnMethodName.Revoke,
+      permissionsGrantId,
+    };
+    const existingRevokesForGrant: PermissionsRevokeMessage[] = await this.messageStore.query(tenant, query) as PermissionsRevokeMessage[];
+
+    // Conflict 409 if the grant already has an older revoke
+    const oldestExistingRevoke = PermissionsRevoke.getOldestRevoke(existingRevokesForGrant);
+    if (oldestExistingRevoke !== undefined && oldestExistingRevoke.descriptor.dateCreated < message.descriptor.dateCreated) {
+      return {
+        status: { code: 409, detail: 'Conflict' }
+      };
+    }
+
+    // Store incoming PermissionsRevoke
+    const indexes: { [key: string]: string } = {
+      interface          : DwnInterfaceName.Permissions,
+      method             : DwnMethodName.Revoke,
+      permissionsGrantId : message.descriptor.permissionsGrantId,
+    };
+    await this.messageStore.put(tenant, message, indexes);
+    await this.eventLog.append(tenant, await Message.getCid(message));
+
+    // Delete existing revokes which are all older than the incoming message
```

Comment:
Related to https://github.com/TBD54566975/dwn-sdk-js/pull/417/files?diff=unified&w=0#r1245751682